lib/engine-universal/src/engine.rs

```diff
@@ -575,38 +579,34 @@ impl UniversalEngineInner {
             CompileError::Resource(format!("could not allocate code memory: {}", e))
         })?;
         let mut code_writer = unsafe {
-            // SAFETY: We just popped out a “free” code memory from an allocator pool.
-            code_memory.writer()
+            // SAFETY: We just popped out an unused code memory from an allocator pool.
+            code_memory.writer()?
         };
 
         let mut allocated_functions = vec![];
         let mut allocated_data_sections = vec![];
         let mut allocated_executable_sections = vec![];
         for func in function_bodies {
-            let offset = code_writer
-                .write_executable(ARCH_FUNCTION_ALIGNMENT, func.body)
-                .expect("TODO");
+            // TODO: this probably should be considered an error around size calculation above and
+            // panic. Can we ensure somehow there isn’t a need to compute size ahead of time?
+            let offset = code_writer.write_executable(ARCH_FUNCTION_ALIGNMENT, func.body)?;
```

Comment:
Not sure I understand the question at the end of this TODO, but I'm with you that this should panic IMO too (and same below)